src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...newState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          if (newState.hasOwnProperty(key)) {
+            delete newState[key];
+          }
+        });
+        break;
+
+      default:
+        break;
+    }
+    statesArray.push(newState);
+
+    currentState = newState;
```

Comment:
You are directly assigning `newState` to `currentState`, which is not an issue here because `newState` is a copy of `currentState`. However, it's important to note that if `newState` were to be modified in place later on, it would affect `currentState` as well. Since `newState` is already being pushed into `statesArray`, this line is redundant and can be removed.